docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1005 @@
+
+### Differentiation transform
+
+The differentiation transform canonicalizes `differentiable_function` instructions: all `differentiable_function` instructions are filled in with JVP/VJP function values.
+
+```swift
+// Before:
+%fn_ref = function_ref @fn : $@convention(thin) (Float) -> Float
+%diff_fn = differentiable_function [wrt 0] %fn_ref
+
+// After differentiation transform:
+%fn_ref = function_ref @fn : $@convention(thin) (Float) -> Float
+%fn_jvp_ref = function_ref @fn_jvp : $@convention(thin) (Float) -> (Float, (Float) -> Float)
+%fn_vjp_ref = function_ref @fn_vjp : $@convention(thin) (Float) -> (Float, (Float) -> Float)
+%diff_fn = differentiable_function [wrt 0] %fn_ref with {%fn_jvp_ref, %fn_vjp_ref}
+// `@fn_jvp` and `@fn_vjp` may be generated.
+```
+
+## The differentiation transform
+
+The differentiation transform is a SIL mandatory pass that implements automatic differentiation. This is the "magic" in the differentiation system that automatically generates derivative functions. This involves a few steps:
+- [**Activity analysis**](#activity-analysis): static analysis that answers what values need a derivative.
+- **Differentiability checking**: errors are produced for non-differentiable operations and warnings are produced for accidental data flow mistakes.
+- **Automatic differentiation**: generate derivative functions.
+
+### Activity analysis[†](https://www-sop.inria.fr/tropics/papers/supportCoursDA.pdf)
+
+Activity analysis is a static SIL data flow analysis that determines exactly what values need a derivative. Let’s walk through the following example:
+
+```swift
+@differentiable
+func f(_ x: Float) -> Float {
+    let result = sin(x) * cos(3)
+    print(result)
+    return result
+}
+```
+
+Activity analysis classifies values in a function into these categories:
+- **Varied values**: values that depend on the input.
+- **Useful values**: values that contribute to the output.
+- **Active values**: values are both varied and useful. These are the values that need a derivative.
+
+Here is the result of activity analysis for this example:
+
+```swift
+// Varied values are in all caps.
+// These values depend on the input (`x`).
+@differentiable
+func f(_ X: Float) -> Float {
+    let SINX = sin(X)
+    let cos3 = cos(3)
+    let RESULT = SINX * cos3
+    print(RESULT)
+    return RESULT
+}
+
+// Useful values are in all caps.
+// These values contribute to the output (`result`).
+@differentiable
+func f(_ X: Float) -> Float {
+    let SINX = sin(X)
+    let COS3 = cos(3) // 3 should be treated as if capitalized.
+    let RESULT = SINX * COS3
+    print(RESULT)
+    return RESULT
+}
+
+// Active values are in all caps.
+// These values are varied and useful, and thus need a derivative.
+@differentiable
+func f(_ X: Float) -> Float {
+    let SINX = sin(X)
+    let cos3 = cos(3)
+    let RESULT = SINX * cos3
+    let void = print(result)
+    return RESULT
+}
```

Comment:
All caps are a bit hard to read. Can you use a pair of `*` to highlight things?  That wouldn't make it compile but it would convey information more clearly for the purpose of this section.

Reply:
```suggestion
// Varied values are surrounded by asterisks.
// These values depend on the input (`x`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    print(**result**)
    return **result**
}

// Useful values are surrounded by asterisks.
// These values contribute to the output (`result`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let **cos3** = cos(**3**)
    let **result** = **sinx** * **cos3**
    print(**result**)
    return **result**
}

// Active values are surrounded by asterisks.
// These values are varied and useful, and thus need a derivative.
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    let void = print(**result**)
    return **result**
}
```

Reply:
```suggestion
// Varied values are surrounded by asterisks.
// These values depend on the input (`x`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    print(**result**)
    return **result**
}

// Useful values are surrounded by asterisks.
// These values contribute to the output (`result`).
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let **cos3** = cos(**3**)
    let **result** = **sinx** * **cos3**
    print(**result**)
    return **result**
}

// Active values are surrounded by asterisks.
// These values are varied and useful, and thus need a derivative.
@differentiable
func f(_ **x**: Float) -> Float {
    let **sinx** = sin(**x**)
    let cos3 = cos(3)
    let **result** = **sinx** * cos3
    let void = print(**result**)
    return **result**
}
```